### ISSUE TITLE
Very bad latency via X11 (tcp) on 1Gb LAN (in comparison to xterm)

### ISSUE THREAD

**Issue by lukolszewski** [alacritty/alacritty#7129]
I'm running an x terminal via X11 on a local gigabit lan (single hop, barely any other use). My x terminal of choice is alacritty thanks to its OSC52 and good unicode support. However, I've noticed horrible latency (worse than chrome running over the same network). The latency is (by feel) on the order of many hundreds of milliseconds for screen redraw. For example scrolling text in vim I have to literally wait for every screen redraw. I just assumed this is a limitation of X11 when used on a large display even though I'm running it over 1Gb network with sub ms latency. However, after I tried using Xterm and it essentially runs instantly(I can;t tell its running remotely) I decided this might be a bug. So here it is.

Perhaps there are some settings that can be tweaked?

### System

OS: Linux
Version:0.12.2
X11 on the host: Arch Linux, i3, X11, Nvidia rtx2070 GPU proprietary driver, 4K screen
X11 on the client: AlmaLinux 9,X11,  GNOME 40, laptop

### Logs

Created log file at "/tmp/Alacritty-22970.log"
[0.000001642s] [INFO ] [alacritty] Welcome to Alacritty
[0.000192300s] [INFO ] [alacritty] Version 0.12.2
[0.000206508s] [INFO ] [alacritty] Running on X11
[0.000635748s] [INFO ] [alacritty] Configuration files loaded from:
                                     "/home/corpo.t-mobile.pl/lolszewski2/.config/alacritty/alacritty.yml"
                                     "/home/corpo.t-mobile.pl/lolszewski2/.config/alacritty/themes/themes/gruvbox_material_medium_dark.yaml"
[0.047501209s] [INFO ] [alacritty] Using GLX 1.4
[0.145723301s] [INFO ] [alacritty] Window scale factor: 1.875
[0.156718424s] [DEBUG] [alacritty] Loading "FiraCode Nerd Font Mono" font
[0.161740370s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.162682940s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.163591005s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.164485538s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.175723930s] [INFO ] [alacritty] Running on llvmpipe (LLVM 14.0.6, 256 bits)
[0.175741079s] [INFO ] [alacritty] Using OpenGL 3.3 renderer
[0.180995407s] [DEBUG] [alacritty] Filling glyph cache with common glyphs
[0.186830004s] [INFO ] [alacritty] Cell size: 22 x 44
[0.186849603s] [INFO ] [alacritty] Padding: 0 x 0
[0.186853607s] [INFO ] [alacritty] Width: 800, Height: 600
[0.209457514s] [INFO ] [alacritty] PTY dimensions: 13 x 36
[0.212281297s] [INFO ] [alacritty] Initialisation complete
[0.280263046s] [DEBUG] [alacritty_terminal] New num_cols is 86 and num_lines is 47
[0.282513506s] [INFO ] [alacritty] Padding: 0 x 0
[0.282578001s] [INFO ] [alacritty] Width: 1912, Height: 2084
[4.231879313s] [INFO ] [alacritty] Goodbye

**Comment by kchibisov** [comment]
Have you tried using vnc instead of X11 forwarding, X11 forwarding is known to be bad with dri stuff?

**Comment by lukolszewski** [comment]
I can't use VNC on this device (I have no root, also security policy prohibits setting up services listening for connectivity so also no SSH). However, xterm runs pretty much as fast as local via X11 on my LAN. Is it DRI that is the differentiator? (xterm = no dri, alacritty = uses dri)?

**Comment by kchibisov** [comment]
Well, xterm is more suited for what you want, it uses X11 to render, so no surprise it works well. You don't need root for vnc, though, I guess you want to stick with X11, and I'm not sure how it works there. On Wayland you can run all rootless without any issue and have good latency.

**Comment by kchibisov** [comment]
>Also, I tried running glxgears the same way(X11 over LAN). I'm getting 7 FPS. I'm not sure if this is relevant

it's dri stuff wasn't meant to run like that.

**Comment by lukolszewski** [comment]
I don't suppose alacritty can run without DRI (with indirect rendering)? I found some software (for example chrome) supports this switch LIBGL_ALWAYS_INDIRECT=1 to force indirect rendering of OpenGL. Supposedly it speeds up its responsiveness using X11 over the network a lot.

I tried it with alacritty, but I got the following error:

Error: [a7] GLXBadFBConfig
Error: "Event loop terminated with code: 1"

So I guess the answer (at least at this stage) is no.
